mindmeld/components/dialogue.py

```diff
@@ -689,6 +692,189 @@ def _get_dialogue_handler(self, dialogue_state):
+
+        For slot filling, the user query will consist of the prompted entity.
+        Hence `entity = [query]` and `entity_index = 0`.
+
+        """
+        formatted_payload = (query, [query], 0)
+
+        if slot.default_eval:
+            if entity_type in DEFAULT_SYS_ENTITIES:
+                # system entity validation - checks for presence of required system entity
+                resources = {}
+                extracted_feature = dict(
+                    query_features.extract_sys_candidates([entity_type])(query, resources)
+                )
+            else:
+                # gazetteer validation
+                gazetteer = (
+                    {'gazetteers':
+                        {entity_type:
+                            self._app.app_manager.nlp.resource_loader.get_gazetteer(entity_type)
+                        }
+                    }
+                )
+                extracted_feature = (
+                entity_features.extract_in_gaz_features()(formatted_payload, gazetteer))
+
+        if slot.hints:
+            # hints / user-list validation
+            extracted_feature = {'hint_validated_entity': text} if text in slot.hints else {}
+
+        # return True iff user input results in extracted features (i.e. validated)
+        return extracted_feature != {}
+
+    def _initial_fill(self, request):
+        """
+        Performs the first pass and fills the entity form with entity values available
+        in the initial query.
+
+        Args:
+            request (Request): The request object.
+        """
+        for entity in request.entities:
+            entity_type = entity['type']
+            role = entity['role']
+            value = entity['text']
+
+            for slot in self._local_form:
+                if entity_type == slot.entity:
+                    if slot.role == None or (role == slot.role):
+                        slot.value = value
+
+    def __call__(self, request, responder):
+        """The iterative call to fill missing slots in the entity form till all slots have been
+        filled up or the flow has been exited.
+
+        Args:
+            request (Request): The request object.
+            responder (DialogueResponder): The responder object.
+        """
+        if self._slot_not_prompted is None or self._local_form is None:
+            # Entering the flow
+            if not self._entity_form:
+                self._exit_flow(responder)
+                return
+
+            self._slot_not_prompted = True
+            # Fill the form with the entities in the first query
+            self._local_form = copy.deepcopy(self._entity_form)
+            self._initial_fill(request)
+
+        for slot in self._local_form:
+            entity_ = slot.entity
+            value = slot.value
+            nlr = (slot.responses if slot.responses
+                else ["Please provide value for: {}".format(entity_)]
```

Comment:
Should we move this into validation or some earlier steps?

Reply:
This is just setting the default nlr at an object level. This is the first time the object is accessed from the form, I don't see where it can go earlier. Validation is a later step. This is just a default user prompt for the slot value, validation is done on the value they return.